compiler/rustc_infer/src/infer/relate/combine.rs

```diff
@@ -178,18 +178,15 @@ impl<'tcx> InferCtxt<'tcx> {
                 relation.param_env().and((a.ty(), b.ty())),
                 &mut OriginalQueryValues::default(),
             );
-            self.tcx.check_tys_might_be_eq(canonical).unwrap_or_else(|_| {
+            self.tcx.check_tys_might_be_eq(canonical).map_err(|_| {
                 // The error will only be reported later. If we emit an ErrorGuaranteed
                 // here, then we will never get to the code that actually emits the error.
                 self.tcx.dcx().delayed_bug(format!(
                     "cannot relate consts of different types (a={a:?}, b={b:?})",
                 ));
-                // We treat these constants as if they were of the same type, so that any
-                // such constants being used in impls make these impls match barring other mismatches.
-                // This helps with diagnostics down the road.
-            });
-        });
-
+                TypeError::Mismatch
+            })
+        })?;
```

Comment:
I wonder if we should actually just do...

```
self.at(...).equate(a.ty(), b.ty())?;
```

...here, rather than reimplementing that behavior in a query that just approximates if types are equal. I don't think I'm convinced about:

```
// This probe is probably not strictly necessary but it seems better to be safe and not accidentally find
// ourselves with a check to find bugs being required for code to compile because it made inference progress.
```

since I believe the only case where this may occur is GCE[^1], and if we make inference-breaking changes to GCE, then like... it's nightly? :laughing:

[^1]: Though I guess we should check that this doesn't interact poorly with associated consts or something, though I'm not certain where we'd ever have a case where equating the types of two consts causes inference guidance that we *don't* want...

Reply:
oh 🤦 yea, at this point we should just equate them

Reply:
Apparently under GCE it does cause something I haven't debugged further, and just turned the test into a known-bug.